tensorpipe/common/cuda_host_allocator.cc

```diff
@@ -14,31 +14,35 @@
 
 namespace tensorpipe {
 
+namespace {
+
+uint8_t* allocPinnedBuffer(size_t size) {
+  uint8_t* ptr;
+  TP_CUDA_CHECK(cudaMallocHost(&ptr, size));
+  return ptr;
+}
+
+void freePinnedBuffer(uint8_t* ptr) {
+  TP_CUDA_CHECK(cudaFreeHost(ptr));
+}
+
+} // namespace
+
 CudaHostAllocator::CudaHostAllocator(size_t numChunks, size_t chunkSize)
     : numChunks_(numChunks),
       chunkSize_(chunkSize),
-      data_(allocPinnedBuffer(numChunks * chunkSize)),
-      chunkAvailable_(numChunks, true) {
-  std::unique_lock<std::mutex> lock(mutex_);
-  thread_ = std::thread([this, lock{std::move(lock)}]() mutable {
-    setThreadName("TP_CUDA_host_allocator_loop");
-    allocLoop(std::move(lock));
-  });
-}
+      data_(allocPinnedBuffer(numChunks * chunkSize), freePinnedBuffer),
+      chunkAvailable_(numChunks, true) {}
 
 CudaHostAllocator::~CudaHostAllocator() {
   join();
 }
 
 void CudaHostAllocator::alloc(size_t size, TAllocCallback callback) {
   std::unique_lock<std::mutex> lock(mutex_);
```

Comment:
Right, so as long as `alloc()`, `close()`, and `releaseChunk()` are always called from the context loop, we can get rid of the mutex. Removing.

Reply:
If you want to make this assumption explicit, hence make it both safer and clearer to the future readers of this code, you could make it so that the constructor of the allocator takes a const ref to a DeferredExecutor, store it, and here call `TP_DCHECK(loop_.inLoop())`.